pkg/model/raid_array.go

```diff
@@ -165,3 +161,106 @@ func (a *RaidArray) DeleteHardware(ctx context.Context) error {
 
 	return VirtualDiskNotFoundError(a)
 }
+
+func ListVirtualDisks(ctx context.Context, raidType string) (virtualDisks []*common.VirtualDisk, err error) {
+	switch raidType {
+	case common.SlugRAIDImplLinuxSoftware:
+		return listVirtualDisksLinux(ctx)
+	case common.SlugRAIDImplHardware:
+		return listVirtualDisksHardware(ctx)
+	default:
+		err = InvalidRaidTypeError(raidType)
+		return
+	}
+}
+
+func ListPhysicalDisks(ctx context.Context, raidType string) (physicalDisks []*common.Drive, err error) {
+	switch raidType {
+	case common.SlugRAIDImplLinuxSoftware:
+		return listPhysicalDisksLinux(ctx)
+	case common.SlugRAIDImplHardware:
+		return listPhysicalDisksHardware(ctx)
+	default:
+		err = InvalidRaidTypeError(raidType)
+		return
+	}
+}
+
+func listVirtualDisksLinux(_ context.Context) (virtualDisks []*common.VirtualDisk, err error) {
+	// TODO(splaspood) Implement VD listing for mdadm
+	// mdadm --misc --detail --export /dev/md/<name>*
+	// Seems on my test ubuntu 20.04 host these end up named /dev/md/ROOT_0 (that _0)
+	return
+}
+
+func listPhysicalDisksLinux(_ context.Context) (physicalDisks []*common.Drive, err error) {
+	// TODO(splaspood) Implement PD listing for mdadm/noraid
+	return
+}
+
+func listVirtualDisksHardware(ctx context.Context) (virtualDisks []*common.VirtualDisk, err error) {
+	hardware, err := getIronlibInventory(ctx)
+
+	for _, sc := range hardware.StorageControllers {
+		if sc.Vendor == common.VendorMarvell {
+			var sca *actions.StorageControllerAction
+			sca, err = getStorageControllerAction(ctx)
+			if err != nil {
+				return
+			}
+
+			virtualDisks, err = sca.ListVirtualDisks(ctx, sc)
+			if err != nil {
+				return
+			}
+		}
+	}
+
+	return
+}
+
+func listPhysicalDisksHardware(ctx context.Context) (physicalDisks []*common.Drive, err error) {
+	hardware, err := getIronlibInventory(ctx)
+
+	for _, sc := range hardware.StorageControllers {
+		if sc.Vendor == common.VendorMarvell {
+			for _, drive := range hardware.Drives {
+				if drive.StorageControllerDriveID >= 0 {
+					physicalDisks = append(physicalDisks, drive)
+				}
+			}
+		}
+	}
+
+	return
+}
+
+func getIronlibInventory(ctx context.Context) (hardware *common.Device, err error) {
+	logrusLogger, err := command.ZapToLogrus(ctx)
+	if err != nil {
+		return
+	}
+
+	device, err := ironlib.New(logrusLogger)
+	if err != nil {
+		return
+	}
+
+	hardware, err = device.GetInventory(ctx, true)
```

Comment:
the extra error check can be skipped here - `return device.GetInventory(ctx, true)`